### PR TITLE
feat: add execution contexts api

### PR DESCRIPTION
**Pull request by zhujian7**
**What this PR does / why we need it**:

Add list execution contexts API to list ExecutionContexts of a tenant.
An execution context describes the context of workflow execution, including information(cluster, namespace, PVC, etc.)that workflows will use in running.

**Which issue(s) this PR is related to** *(optional, link to 3rd issue(s))*:

Fixes #

Reference to #

**Special notes for your reviewer**:

/cc @supereagle @cd1989 

**Release note**:

```release-note
Add execution contexts API
```

### CODE REVIEW

File: pkg/server/apis/v1alpha1/descriptors/tenant.go [hunk 1]
```diff
@@ -88,4 +88,27 @@ var tenant = []definition.Descriptor{
 			},
 		},
 	},
+	{
+		Path:        "/tenants/{tenant}/precheck",
```
Comment (cd1989) [inline pkg/server/apis/v1alpha1/descriptors/tenant.go]: "/tenants/{tenant}/executioncontexts" to list all execution contexts for the given tenant.

File: pkg/server/handler/v1alpha1/precheck.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+			}
+		}
+	}
+
+	return workersContext, nil
+}
+
+// WorkerContextStatus describes the status of worker clusters, it contains information that affects
+// pipeline execution, like reserved resources, pvc status.
+type WorkerContextStatus struct {
```
Comment (cd1989) [inline pkg/server/handler/v1alpha1/precheck.go]:
```
ExecutionContext {
  Spec: ...
  Status: ...
}
```

File: pkg/server/apis/v1alpha1/descriptors/tenant.go (hunk 1 above)
Comment (zhujian7) [reply pkg/server/apis/v1alpha1/descriptors/tenant.go]: done

File: pkg/server/handler/v1alpha1/precheck.go (hunk 2 above)
Comment (zhujian7) [reply pkg/server/handler/v1alpha1/precheck.go]: fixed

**Comment by zhujian7** [comment]
@cd1989 PTAL

File: pkg/server/apis/v1alpha1/types.go [hunk 3]
```diff
@@ -248,3 +248,40 @@ type StorageUsage struct {
 type StorageCleanup struct {
 	Paths []string `json:"paths"`
 }
+
+// ExecutionContextSpec describes the execution context
+type ExecutionContextSpec struct {
+	Cluster     string `json:"cluster"`
+	Namespace   string `json:"namespace"`
+	Integration string `json:"integration"`
```
Comment (cd1989) [inline pkg/server/apis/v1alpha1/types.go]: Do we need `integration` ?

File: pkg/server/apis/v1alpha1/types.go [hunk 4]
```diff
@@ -248,3 +248,40 @@ type StorageUsage struct {
...
+	// Phase of the execution context, could be 'Ready', 'NotReady' or 'Unknown'
+	Phase             ExecutionContextPhase                `json:"phase"`
+	ReservedResources map[core_v1.ResourceName]string      `json:"reservedResources"`
+	PVC               *core_v1.PersistentVolumeClaimStatus `json:"pvc"`
+}
+
+// ExecutionContext represtents a context used to execute workflows.
+type ExecutionContext struct {
+	// Metadata for the particular object, including name, namespace, labels, etc
+	meta_v1.ObjectMeta `json:"metadata,omitempty"`
```
Comment (cd1989) [inline pkg/server/apis/v1alpha1/types.go]: I think ExecutionContext doesn't have concept like `namespace`, and also no need to use `labels`, it's not k8s resources, and we should not treat like that.

File: pkg/server/apis/v1alpha1/types.go [hunk 5]
```diff
@@ -248,3 +248,40 @@ type StorageUsage struct {
...
+	Integration string `json:"integration"`
+	PVC         string `json:"pvc"`
+}
+
+// ExecutionContextStatus describe the status of execution context, it contains information that affects
+// pipeline execution, like reserved resources, pvc status.
+type ExecutionContextStatus struct {
+	// Phase of the execution context, could be 'Ready', 'NotReady' or 'Unknown'
+	Phase             ExecutionContextPhase                `json:"phase"`
+	ReservedResources map[core_v1.ResourceName]string      `json:"reservedResources"`
```
Comment (cd1989) [inline pkg/server/apis/v1alpha1/types.go]: You can comment what is ReservedResources with examples.

File: pkg/server/apis/v1alpha1/types.go (hunk 4 above)
Comment (zhujian7) [reply pkg/server/apis/v1alpha1/types.go]: OK，I will delete it.

File: pkg/server/apis/v1alpha1/types.go (hunk 3 above)
Comment (zhujian7) [reply pkg/server/apis/v1alpha1/types.go]: I will add a `Closed` phase for execution context.
If users want to use a closed execution context, they need to open the related cluster integration firstly. 
So I think the `integration` field is useful.

File: pkg/server/apis/v1alpha1/types.go (hunk 5 above)
Comment (zhujian7) [reply pkg/server/apis/v1alpha1/types.go]: OK

**Comment by zhujian7** [comment]
@cd1989 PTAL

File: pkg/server/apis/v1alpha1/types.go (hunk 3 above)
Comment (cd1989) [reply pkg/server/apis/v1alpha1/types.go]: I'd prefer to remove it.

**Pull request merged**